### PR TITLE
[OB-17] shipcat shell will no longer attempt to connect to broken pods

### PR DESCRIPTION
**Pull request by RhysStansfield**
`shipcat shell` will no longer attempt to connect to broken pods.

Aborts if no healthy pods.

### CODE REVIEW

File: shipcat_cli/src/kubectl.rs [hunk 1]
```diff
@@ -145,25 +147,32 @@ fn get_broken_pods(mf: &Manifest) -> Result<(String, Vec<String>)> {
...
+    let mut results = HashMap::new();
+    results.insert("active", vec![]);
+    results.insert("broken", vec![]);
+
     for l in podres.lines() {
         if !l.contains("Running") {
             if let Some(p) = l.split(' ').next() {
                 warn!("Found pod not running: {}", p);
-                bpods.push(p.into());
+                results.get_mut("broken").unwrap().push(p.into());
```
Comment (clux) [inline shipcat_cli/src/kubectl.rs]: maybe broken map and active map mut at start

File: shipcat_cli/src/kubectl.rs [hunk 2]
```diff
@@ -396,6 +405,11 @@ pub fn kpods(mf: &Manifest) -> Result<String> {
 pub fn shell(mf: &Manifest, desiredpod: Option<usize>, cmd: Option<Vec<&str>>) -> Result<()> {
     // TODO: kubectl auth can-i create pods/exec
     let podsres = get_pods(&mf)?;
+    debug!("podsres: {}", podsres);
+    if podsres.is_empty() {
+        warn!("No healthy pods for {}, cannot connect", mf.name);
```
Comment (clux) [inline shipcat_cli/src/kubectl.rs]: bail!

File: shipcat_cli/src/kubectl.rs [hunk 3]
```diff
@@ -120,23 +121,24 @@ pub fn await_rollout_status(mf: &Manifest) -> Result<bool> {
...
     Ok(podsres)
 }
 
 /// Return non-running or partially ready pods
 fn get_broken_pods(mf: &Manifest) -> Result<(String, Vec<String>)> {
+    let (podres, podmap) = segregate_pods(mf).unwrap();
+    Ok((podres, podmap.get("broken").unwrap().clone()))
+}
+
+fn segregate_pods(mf: &Manifest) -> Result<(String, HashMap<&str, Vec<String>>)> {
```
Comment (clux) [inline shipcat_cli/src/kubectl.rs]: get_classified_pods (mf, PodState) -> REsult<String,Vec>

**Comment by janslow** [comment]
Isn't that a good thing?

File: shipcat_cli/src/kubectl.rs [hunk 4]
```diff
@@ -120,23 +121,24 @@ pub fn await_rollout_status(mf: &Manifest) -> Result<bool> {
...
-        format!("-l=app={}", mf.name),
-        format!("-n={}", mf.namespace),
-        "-o".into(),
-        "jsonpath='{.items[*].metadata.name}'".into(),
-    ];
-    // TODO: filter out ones not running conditionally - exec wont work with this
-    let (podsres, _) = kout(podargs)?;
+    let (_, podmap) = get_classified_pods(mf).unwrap();
+    let mut podsres : String;
+    if let Some(active) = podmap.get("active") {
```
Comment (clux) [inline shipcat_cli/src/kubectl.rs]: `let podres = if let...` and return an expression

File: shipcat_cli/src/kubectl.rs [hunk 5]
```diff
@@ -145,25 +147,36 @@ fn get_broken_pods(mf: &Manifest) -> Result<(String, Vec<String>)> {
...
             }
         }
     }
-    Ok((podres, bpods))
+
+    let mut results = HashMap::new();
+    results.insert("active", active);
+    results.insert("broken", broken);
+
+    Ok((podres, results))
```
Comment (clux) [inline shipcat_cli/src/kubectl.rs]: and return either active or broken as the second part of the tuple

File: shipcat_cli/src/kubectl.rs [hunk 6]
```diff
@@ -145,25 +143,35 @@ fn get_broken_pods(mf: &Manifest) -> Result<(String, Vec<String>)> {
         "--no-headers".into(),
     ];
     let (podres, _) = kout(podargs)?;
-    let mut bpods = vec![];
     let status_re = Regex::new(r" (?P<ready>\d+)/(?P<total>\d+) ").unwrap();
+    let mut active = vec![];
+    let mut broken = vec![];
+
     for l in podres.lines() {
         if !l.contains("Running") {
```
Comment (janslow) [inline shipcat_cli/src/kubectl.rs]: Does this also detect `Terminating` pods?

Comment (clux) [reply shipcat_cli/src/kubectl.rs]: Terminating is mutually exclusive with Running here.

**Comment by RhysStansfield** [comment]
@janslow Trying to connect to a broken pod was not a good thing, so this prevents that from happening!

**Comment by janslow** [comment]
> @janslow Trying to connect to a broken pod was not a good thing, so this prevents that from happening!

I thought this was an issue complaining that Shipcat doesn't connect to them originally.

FYI, this probably closes https://github.com/Babylonpartners/shipcat/issues/244

**Comment by RhysStansfield** [comment]
@janslow Yeah I could have worded this better, and yes this will also resolve #244

**Pull request merged**